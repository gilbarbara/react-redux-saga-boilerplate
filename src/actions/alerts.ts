import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
