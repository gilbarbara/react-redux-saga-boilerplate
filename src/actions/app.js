// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
