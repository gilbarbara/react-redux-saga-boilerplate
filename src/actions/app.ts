/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
