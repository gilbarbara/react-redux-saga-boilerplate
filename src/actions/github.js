// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { githubGetRepos: getRepos } = createActions({
  [ActionTypes.GITHUB_GET_REPOS]: (query: string) => ({ query }),
});
