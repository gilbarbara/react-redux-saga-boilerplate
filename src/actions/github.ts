/**
 * @module Actions/User
 * @desc User Actions
 */
import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

export const getRepos = createAction(ActionTypes.GITHUB_GET_REPOS_REQUEST, (query: string) => ({
  query,
}));
