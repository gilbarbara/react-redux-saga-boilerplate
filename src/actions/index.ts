export * from './app';
export * from './github';
export * from './user';
