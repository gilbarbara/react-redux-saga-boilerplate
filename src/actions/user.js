// @flow
/**
 * @module Actions/User
 * @desc User Actions
 */
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export const { userLogin: login, userLogout: logOut } = createActions({
  [ActionTypes.USER_LOGIN]: () => ({}),
  [ActionTypes.USER_LOGOUT]: () => ({}),
});
