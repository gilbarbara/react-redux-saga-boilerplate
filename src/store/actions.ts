export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
