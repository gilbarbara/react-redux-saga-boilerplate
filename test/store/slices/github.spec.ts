import { getRepos, getReposFailure, getReposSuccess } from '~/actions';

import reducer from '~/store/slices/github';

describe('slices/github', () => {
  describe('actions', () => {
    it('getRepos', () => {
      expect(getRepos('react')).toMatchSnapshot();
    });
  });

  describe('reducer', () => {
    it('should return the initial state', () => {
      expect(reducer(undefined, { type: '', payload: {} })).toMatchSnapshot();
    });

    describe('getRepos', () => {
      it('should handle REQUEST', () => {
        expect(reducer(undefined, getRepos('react'))).toMatchSnapshot();
      });

      it('should handle SUCCESS', () => {
        const initialState = reducer(undefined, getRepos('react'));

        expect(
          reducer(initialState, {
            type: getReposSuccess.type,
            payload: {
              data: [{ name: 'one' }],
            },
            meta: {
              query: 'react',
              updatedAt: 1234567890,
            },
          }),
        ).toMatchSnapshot();
      });

      it('should handle FAILURE', () => {
        const initialState = reducer(undefined, getRepos('react'));

        expect(
          reducer(initialState, {
            type: getReposFailure.type,
            payload: {
              message: 'Something failed',
            },
            meta: {
              query: 'react',
            },
          }),
        ).toMatchSnapshot();
      });
    });
  });
});
