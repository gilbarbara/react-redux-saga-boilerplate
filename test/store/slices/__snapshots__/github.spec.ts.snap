// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`slices/github > actions > getRepos 1`] = `
{
  "payload": "react",
  "type": "github/getRepos",
}
`;

exports[`slices/github > reducer > getRepos > should handle FAILURE 1`] = `
{
  "topics": {
    "react": {
      "cached": false,
      "data": [],
      "message": {
        "message": "Something failed",
      },
      "status": "error",
      "updatedAt": 0,
    },
  },
}
`;

exports[`slices/github > reducer > getRepos > should handle REQUEST 1`] = `
{
  "topics": {
    "react": {
      "cached": false,
      "data": [],
      "message": "",
      "status": "running",
      "updatedAt": 0,
    },
  },
}
`;

exports[`slices/github > reducer > getRepos > should handle SUCCESS 1`] = `
{
  "topics": {
    "react": {
      "cached": false,
      "data": {
        "data": [
          {
            "name": "one",
          },
        ],
      },
      "message": "",
      "status": "success",
      "updatedAt": 1234567890,
    },
  },
}
`;

exports[`slices/github > reducer > should return the initial state 1`] = `
{
  "topics": {},
}
`;
